Refactor intermediate datatypes to make the compiler output stable

mcintyre94 opened the pull request:
This means that if we compile the same code repeatedly the output is identical. The motivation for this is to allow adding lightweight infrastructure to detect whether/how a PR changes the compile output for the example Python files. Having a stable output is a prerequisite for this to avoid false positives where eg. functions are just rearranged.

Where we've used `HashMap` and `HashSet` data structures the sort order is undefined, and in practice if you compile multiple times you'll get different outputs. 

See suggestion in Discord: https://discordapp.com/channels/1005658120548270224/1006027721127776388/1042345155522474034

To test this I compiled the examples 20 times each:

```sh
# examples/
for f in ./*.py
do
    for i in {1..20}
    do
        ../target/debug/seahorse compile $f > compiled/$f$i.rs
    done
done
```

And then verified with `md5 compiled/*` that all instances of the same file are identical

Intentionally didn't update changelog BTW! :D

ameliatastic commented:
Alternate minimal-effort idea: replace every HashSet/HashMap with a BTreeSet/BTreeMap. Assuming we're not using any hash-specific functionality, this might just work

mcintyre94 commented:
> Alternate minimal-effort idea: replace every HashSet/HashMap with a BTreeSet/BTreeMap. Assuming we're not using any hash-specific functionality, this might just work

TIL about those, thanks! I'll give that a go, I think it'd be better to keep the insertion order than sort by name if possible. And hopefully a bit more straightforward too :)

ameliatastic commented:
> I think it'd be better to keep the insertion order than sort by name if possible

Could be done with yet another type! I know that Python does that, [this post](https://medium.com/junior-dev/python-dictionaries-are-ordered-now-but-how-and-why-5d5a40ee327f) explains how (hopefully it's correct lol, just picked the first one I found). I think I like name order slightly better though, one advantage is that you have some flexibility with changing how things are added without accidentally changing the output. What draws you to insertion order?

mcintyre94 commented:
> I think I like name order slightly better though, one advantage is that you have some flexibility with changing how things are added without accidentally changing the output. What draws you to insertion order?

Ohh I was assuming `BTreeMap` etc were using insertion order. Reasoning was just that it'd be closer in structure to the input. I'm not strongly opinionated on the output order though.

I've updated the PR to use those data structured, think that's a lot better! Same test I used before worked! I just needed to update `inferred_accounts` and `namespace`. I also updated `features` kinda speculatively, it won't affect any current examples but I think it would affect output if we had multiple features.
